In [1]:
import pandas as pd

# 1. Import des données

In [ ]:
df_irve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/eb76d20a-8501-400e-b336-d85724de5435'
)

print(f"Shape : {df_irve.shape}")
df_irve.sample(5)

Shape : (212234, 52)


C:\Users\lauri\AppData\Local\Temp\ipykernel_51852\2674767103.py:1: DtypeWarning: Columns (12,18,19,20,21,22,24,25,29,33) have mixed types. Specify dtype option on import or set low_memory=False.
  df_irve = pd.read_csv(


,nom_amenageur,siren_amenageur,contact_amenageur,nom_operateur,contact_operateur,telephone_operateur,nom_enseigne,id_station_itinerance,id_station_local,nom_station,...,datagouv_resource_id,datagouv_organization_or_owner,created_at,consolidated_longitude,consolidated_latitude,consolidated_code_postal,consolidated_commune,consolidated_is_lon_lat_correct,consolidated_is_code_insee_verified,consolidated_is_code_insee_modified
21643,Métropole Rouen Normandie,NaN,NaN,Bouygues Energies & Services,support@alizecharge.fr,0805021480,Métropole Rouen Normandie,FRC01E76069001,NaN,BELBEUF - Rue Des Canadiens,...,5f29737c-5393-46f9-8140-2509992adc7a,alize,2025-03-17T08:58:10.352000+00:00,1.140311,49.388914,76240.0,Belbeuf,True,True,False


Ce premier jeu de données est la base nationale des IRVE (Infrastructures de Recharge pour Véhicules Électriques).
Cette table comporte 212234 lignes pour 52 variables.

In [4]:
df_ve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/90e0d717-deda-4bdc-9987-f82faac5bc93',
 encoding='latin-1'
)

print(f"Shape : {df_ve.shape}")
df_ve.sample(5)

Shape : (703545, 8)


C:\Users\lauri\AppData\Local\Temp\ipykernel_51852\2115914144.py:1: DtypeWarning: Columns (0,2) have mixed types. Specify dtype option on import or set low_memory=False.
  df_ve = pd.read_csv(


,CODGEO,LIBGEO,EPCI,LIBEPCI,DATE_ARRETE,NB_VP_RECHARGEABLES_EL,NB_VP_RECHARGEABLES_GAZ,NB_VP
361293,31168,EOUX,200072643,CC Cœur et Coteaux du Comminges,2023-12-31,4,0,144
165430,51491,SAINT-JEAN-SUR-TOURBE,200042620.0,CC de la Région de Suippes,2024-06-30,0,0,77
616328,44100,MONNIÈRES,200067635.0,CA Clisson Sèvre et Maine Agglo,2024-12-31,55,0,2323
416362,76087,BERVILLE-EN-CAUX,200069847,CC Plateau de Caux-Doudeville-Yerville,2021-09-30,5,0,704
182096,62621,NORTKERQUE,246200844.0,CC de la Région d'Audruicq,2022-09-30,20,0,1861


Ce second jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge. Il contient l'historique des immatriculations pour chaque commune à différentes dates.
Cette table comporte 703545 lignes pour 8 variables. 

In [5]:
df_revenus = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/516130bc-4dcb-47f5-8347-ae96553c43ab',sep=';'
)

print(f"Shape : {df_revenus.shape}")
df_revenus.sample(5)

Shape : (34926, 57)


C:\Users\lauri\AppData\Local\Temp\ipykernel_51852\3484954150.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df_revenus = pd.read_csv(


,Nom géographique GMS,Code géographique,Libellé géographique,[DISP] Nbre de ménages fiscaux,[DISP] Nbre de personnes dans les ménages fiscaux,[DISP] Nbre d'unités de consommation dans les ménages fiscaux,[DISP] 1ᵉʳ quartile (€),[DISP] Médiane (€),[DISP] 3ᵉ quartile (€),[DISP] Écart interquartile (€),...,[DEC] 9ᵉ décile (€),[DEC] Rapport interdécile D9/D1,[DEC] S80/S20,[DEC] Iice de Gini,[DEC] Part des revenus d’activité (%),[DEC] dont part des salaires et traitements (%),[DEC] dont part des iemnités de chômage (%),[DEC] dont part des revenus des activités non salariées (%),"[DEC] Part des pensions, retraites et rentes (%)",[DEC] Part des autres revenus (%)
13146,Saussines,34296,Saussines,410.0,999.0,674.3,NaN,25470.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4332,Gardanne,13041,Gardanne,9432.0,20865.0,14466.6,17140.0,22780.0,29460.0,12330.0,...,40400.0,"4,5",5.7,0.305,69.7,63.7,3.1,2.9,25.9,4.4
6875,Fussey,21289,Fussey,49.0,116.0,79.3,NaN,23340.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
34136,Saint-Aubin-sur-Yonne,89335,Saint-Aubin-sur-Yonne,196.0,424.0,298.9,NaN,22870.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
20565,Cléguérec,56041,Cléguérec,1225.0,2705.0,1890.6,17210.0,22080.0,27500.0,10290.0,...,36150.0,"3,1",4.2,0.280,63.2,51.2,2.3,9.7,31.2,5.6


Ce dernier jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge.
Cette table comporte 34926 lignes pour 57 variables.

# 2. Jointure des tables

## 2.1 Choix de la variable de jointure

Les 3 jeu de données contiennent une variable indiquant le code géographique mais sous 3 noms différents :
- `code_insee_commune` pour df_irve
- `CODGEO` pour df_ve
- `Code géographique` pour df_revenus

La jointure sera faite sur ces variables.

## 2.2 Préparation de la variable de jointure

Le diagnostic met en évidence plusieurs points d’attention :
- Présence de valeurs manquantes dans `code_insee_commune` (df_irve)  
- Présence de longueurs de codes incohérentes :
  - `code_insee_commune` (df_irve) contient des codes de longueur 3, 6 et 7
  - `CODGEO` (df_ve) contient des codes de longueur 4
- Les variables de jointure ne sont pas uniques :
  - `code_insee_commune` (df_irve)
  - `CODGEO` (df_ve)

Les clés de jointure présentent des problèmes de qualité et de format.  
Il est nécessaire de les nettoyer et les uniformiser (type, format, gestion des valeurs manquantes, unicité) avant d’effectuer la jointure.

### 2.2.1 Unification des codes

In [ ]:
from src.preparation_data import nettoyer_code_insee

In [ ]:
df_irve["code_insee_commune"] = df_irve["code_insee_commune"].apply(nettoyer_code_insee)
df_ve["CODGEO"] = df_ve["CODGEO"].apply(nettoyer_code_insee)
df_revenus["Code géographique"] = df_revenus["Code géographique"].apply(nettoyer_code_insee)

Les codes insee sont maintenant au même format dans les 3 bases de données : un code à 5 caractères.

### 2.2.2 Compléter les valeurs manquantes

In [ ]:
from src.preparation_data import creer_gdf_irve, charger_communes, joindre_communes, ajouter_codes_geo

In [ ]:
gdf_irve = creer_gdf_irve(df_irve, "consolidated_longitude", "consolidated_latitude")
gdf_communes = charger_communes()
gdf_result = joindre_communes(gdf_irve, gdf_communes)
df_irve = ajouter_codes_geo(df_irve, gdf_result)

Nous avons maintenant 2 colonnes concernant le code géographique dans la base de données IRVE :
- `code_insee_commune` : la variable initiale
- `code_geo_total` : la variable recalculant tous les codes géographiques à partir de la latitude et la longitude

In [ ]:
from src.preparation_data import compter_valeurs_manquantes, compter_uniques

In [ ]:
colonnes_irve = ["code_insee_commune", "code_geo_total"]

print("Valeurs manquantes :")
print(compter_valeurs_manquantes(df_irve, colonnes_irve))

print("\nValeurs uniques :")
print(compter_uniques(df_irve, colonnes_irve))

Cette manipulation permet de passer de 57919 valeurs manquantes à 1768. De plus les codes recalculés sont plus fiables que ceux initiaux.

## 2.3 Jointure

Réfléchissons aux variables à garder... Lesquelles pourraient être intéressante ?

In [6]:
list(df_irve.columns)

['nom_amenageur',
 'siren_amenageur',
 'contact_amenageur',
 'nom_operateur',
 'contact_operateur',
 'telephone_operateur',
 'nom_enseigne',
 'id_station_itinerance',
 'id_station_local',
 'nom_station',
 'implantation_station',
 'adresse_station',
 'code_insee_commune',
 'coordonneesXY',
 'nbre_pdc',
 'id_pdc_itinerance',
 'id_pdc_local',
 'puissance_nominale',
 'prise_type_ef',
 'prise_type_2',
 'prise_type_combo_ccs',
 'prise_type_chademo',
 'prise_type_autre',
 'gratuit',
 'paiement_acte',
 'paiement_cb',
 'paiement_autre',
 'tarification',
 'condition_acces',
 'reservation',
 'horaires',
 'accessibilite_pmr',
 'restriction_gabarit',
 'station_deux_roues',
 'raccordement',
 'num_pdl',
 'date_mise_en_service',
 'observations',
 'date_maj',
 'cable_t2_attache',
 'last_modified',
 'datagouv_dataset_id',
 'datagouv_resource_id',
 'datagouv_organization_or_owner',
 'created_at',
 'consolidated_longitude',
 'consolidated_latitude',
 'consolidated_code_postal',
 'consolidated_commune',
 '

In [7]:
list(df_ve.columns)

['CODGEO',
 'LIBGEO',
 'EPCI',
 'LIBEPCI',
 'DATE_ARRETE',
 'NB_VP_RECHARGEABLES_EL',
 'NB_VP_RECHARGEABLES_GAZ',
 'NB_VP']

In [8]:
list(df_revenus.columns)

['Nom géographique GMS',
 'Code géographique',
 'Libellé géographique',
 '[DISP] Nbre de ménages fiscaux',
 '[DISP] Nbre de personnes dans les ménages fiscaux',
 "[DISP] Nbre d'unités de consommation dans les ménages fiscaux",
 '[DISP] 1ᵉʳ quartile (€)',
 '[DISP] Médiane (€)',
 '[DISP] 3ᵉ quartile (€)',
 '[DISP] Écart interquartile (€)',
 '[DISP] 1ᵉʳ décile (€)',
 '[DISP] 2ᵉ décile (€)',
 '[DISP]3ᵉ décile (€)',
 '[DISP] 4ᵉ décile (€)',
 '[DISP] 6ᵉ décile (€)',
 '[DISP] 7ᵉ décile (€)',
 '[DISP] 8ᵉ décile (€)',
 '[DISP] 9ᵉ décile (€)',
 '[DISP] Rapport interdécile 9ᵉ décile/1ᵉʳ decile',
 '[DISP] S80/20',
 '[DISP] Iice de Gini',
 '[DISP] Part des revenus d’activité (%)',
 '[DISP] dont part des salaires et traitements(%)',
 '[DISP] dont part des iemnités de chômage (%)',
 '[DISP] dont part des revenus des activités non salariées (%)',
 '[DISP] Part des pensions, retraites et rentes (%)',
 '[DISP] Part des revenus du patrimoine et autres revenus (%)',
 "[DISP] Part de l'ensemble des prestat

### 2.3.1 Une ligne par code géo

Il faut réfléchir à la façon dont on veut agréger df_irve et quelles variables on souhaite garder.

##### df_irve

Ici, l'objectif est de mesurer l'offre de recharge par commune. Comme il peut y avoir plusieurs points de recharge par commune, il faut agréger.

**Variables à garder :** nbre_pdc, puissance_nominale, gratuit, paiement_cb, condition_acces.

**Traitement (Aggregation) :**
- nb_stations : Compter le nombre de lignes (ou id_station_local uniques).
- total_pdc : Somme de nbre_pdc (nombre total de points de charge).
- puissance_max : Maximum de puissance_nominale (pour identifier les communes avec recharge ultra-rapide).
- part_gratuit : Pourcentage de points de charge où gratuit est vrai. C'est un excellent indicateur pour l'attractivité d'une commune.
- acces_public : Compter les stations dont la condition_acces est "Accès libre".

In [ ]:
# Nombre de points de recharge par commune
stats_irve_com = df_irve.groupby('code_geo_total').size().reset_index(name='nb_points_recharge')

# puissance totale disponible par commune
stats_puissance_com = df_irve.groupby('code_geo_total')['puissance_nominale'].sum().reset_index(name='puissance_totale_kw')

# Fusion des deux métriques IRVE
df_irve_final = pd.merge(stats_irve_com, stats_puissance_com, on='code_geo_total')
df_irve_final.head()

##### df_ve

Pour que l'analyse soit cohérente avec les données de bornes (IRVE) et de l'INSEE (qui sont des "clichés" à un instant T), il faut isoler la situation la plus récente pour chaque commune afin d'avoir une unique ligne par code insee.

In [ ]:
df_ve["DATE_ARRETE"].dtype

In [ ]:
ve = df_ve.copy()

# Conversion en format date pour être sûr du tri
ve['DATE_ARRETE'] = pd.to_datetime(ve['DATE_ARRETE'])

# Tri par commune puis par date (de la plus ancienne à la plus récente)
ve = ve.sort_values(['CODGEO', 'DATE_ARRETE'])

# garde que la DERNIÈRE ligne pour chaque CODGEO (la plus récente)
df_ve_latest = ve.drop_duplicates(subset='CODGEO', keep='last')

df_ve_latest

**Variables à garder :** NB_VP_RECHARGEABLES_EL, NB_VP.

**Traitement :**
taux_equipement_ve : Créer un ratio (NB_VP_RECHARGEABLES_EL / NB_VP). C'est beaucoup plus pertinent que le chiffre brut, car cela normalise par la taille de la commune.

##### df_revenus

In [ ]:
# On garde le code commune et le revenu médian (souvent colonne 'MED21' pour 2021)
df_revenus_final = df_revenus[['COM', 'MED21']].copy()
df_revenus_final.rename(columns={'COM': 'code_commune_insee', 'MED21': 'revenu_median'}, inplace=True)

**Variables à garder :**
[DISP] Médiane (€), [DISP] Part des revenus d’activité (%), [DISP] Iice de Gini.

Intérêt : La médiane du revenu est indispensable pour vérifier si l'adoption du VE est corrélée à la richesse de la commune. L'indice de Gini peut montrer si les communes très inégalitaires ont un comportement différent.

On peut en choisir d'autres. D'ailleurs on doit choisir uniquement les variables de df_irve avant la jointure car pour les 2 autres tables on a déjà une ligne par commune donc pas besoin de faire d'agrégation. Ainsi on peut toutes les prendre et choisir après.

Après le choix des variables : 
0. variable cible = taux de VE
1. faire des analyses descriptives : 
- automatiser avec une fonction
- matrice de corrélation (heatmap)
2. modélisation :
simple regression ou plus avace (random forest...)